modules/node-shared/src/main/scala/org/tessellation/node/shared/domain/seedlist/SnapshotOrdinalTimeline.scala

```diff
@@ -0,0 +1,37 @@
+package org.tessellation.node.shared.domain.seedlist
+
+import cats.data.{NonEmptyList, Validated}
+import cats.syntax.all._
+
+import org.tessellation.node.shared.domain.seedlist.Range.{mergeOverlaps, parseInput}
+
+import fs2.data.csv.{CellDecoder, DecoderError}
+import io.estatico.newtype.macros.newtype
+
+object SnapshotOrdinalTimeline {
+
+  @newtype
+  case class SnapshotOrdinalTimeline(values: NonEmptyList[Range])
+
+  implicit val optionalSnapshotOrdinalTimelineCellDecoder: CellDecoder[Option[SnapshotOrdinalTimeline]] =
+    CellDecoder.stringDecoder.emap { s =>
+      fromString(s)
+        .leftMap(err => new DecoderError(err.getMessage))
+        .toEither
+    }
+
+  def fromString(str: String): Validated[Throwable, Option[SnapshotOrdinalTimeline]] =
+    parseInput(str)
+      .map(mergeOverlaps)
+      .andThen(_.toNel match {
+        case Some(value) => value.validNel
+        case None        => InvalidInput(str).invalidNel
```

Comment:
I think the `InvalidInput` should be returned by `parseInput`. you can create a different type of error for `SnapshotOrdinalTimeline` if you want, but it looks like you mix two separate things/places together.
Maybe it's enough to return `Either[String, SnapshotOrdinalTimeline]` in here? If there is no string, then you don't need to execute `fromString`, so the optional type here can be avoided I think

Reply:
This function has been changed to

```scala
  def decode(input: String): Validated[NonEmptyList[Throwable], SnapshotOrdinalTimeline] =
    ranges
      .decode(input)
      .map(mergeOverlaps)
      .map(r => NonEmptyList(r.head, r.tail))
      .map(SnapshotOrdinalTimeline(_))
```

If the input is empty then [this error](https://github.com/cngo-github/tessellation/blob/20231228-prot-547/modules/node-shared/src/main/scala/org/tessellation/node/shared/domain/seedlist/Range.scala#L87) will be set on the `ValidationResult`. So, the `List[Range]` return from `mergeOverlaps` will always be non-empty. So, I am converting it to a `NonEmptyList[Range]` unsafely [here](https://github.com/cngo-github/tessellation/blob/20231228-prot-547/modules/node-shared/src/main/scala/org/tessellation/node/shared/domain/seedlist/snapshotOrdinalTimeline.scala#L32).

Reply:
> So, the List[Range] returned from ranges.decode and mergeOverlaps will always be non-empty.

If so, you should return `NonEmptyList[Range]` to ensure in types it will always be non-empty. Unsafe NonEmptyList in here is in general unsafe and relies only on assumption. At some point in the future you may change something in the algorithm and you wouldn't notice the `List[Range]` may return an empty list, so we would introduce a bug in here.

Reply:
Updated the return types of `ranges.decode` and `mergeOverlaps` to be `NonEmptyList` instead of `List`.